compiler/pipes/final-check.cpp

```diff
@@ -587,3 +584,35 @@ void FinalCheckPass::raise_error_using_Unknown_type(VertexPtr v) {
     kphp_error(0, "Using Unknown type");
   }
 }
+
+void FinalCheckPass::check_magic_methods(FunctionPtr fun) {
+  if (!fun->modifiers.is_instance()) {
+    return;
+  }
+
+  const auto name = fun->local_name();
+
+  if (name == ClassData::NAME_OF_TO_STRING) {
+    check_magic_tostring_method(fun);
+  } else if (name == ClassData::NAME_OF_CLONE) {
+    check_magic_clone_method(fun);
+  }
+}
+
+void FinalCheckPass::check_magic_tostring_method(FunctionPtr fun) {
+  const auto count_args = fun->param_ids.size();
+  kphp_error(count_args == 1, fmt_format("Magic method {} cannot take arguments", fun->get_human_readable_name()));
+
+  const auto *ret_type = tinf::get_type(fun, -1);
+  if (!ret_type) {
+    return;
+  }
+
+  kphp_error(ret_type->ptype() == tp_string && ret_type->flags() == 0,
+             fmt_format("Magic method {} must have string return type", fun->get_human_readable_name()));
+}
+
+void FinalCheckPass::check_magic_clone_method(FunctionPtr fun) {
+  kphp_error(!fun->is_resumable, fmt_format("{} method has to be not resumable", ClassData::NAME_OF_CLONE));
+  kphp_error(!fun->can_throw(), fmt_format("{} method should not throw exception", ClassData::NAME_OF_CLONE));
```

Comment:
Let's use `fun->get_human_readable_name()` here instead of `NAME_OF_CLONE`

Reply:
Fixed.